##### Different bacterial and viral pathogens trigger distinct immune responses in a globally invasive ant
##### Supplementary information
##########################################################################################################

The following folder contains all supporting information for the article 10.1038/s41598-019-41843-5.
Clean RNA-seq reads can be accessed from the NCBI SRA server, BioProject ID PRJNA528845 (http://www.ncbi.nlm.nih.gov/bioproject/528845). 

01-data contains:
	
	Immune_qPCR.csv	RT-qPCR
		results for immune genes and pathogens
	gene_count_matrix.csv
		output of StringTie containing raw counts for all Argentine ant transcripts
	genes.TMM.EXPR.matrix
		output of Trinity containing the TMM-normalised counts for non-Argentine ant transcripts
	lhum_immune_genes.csv
		long-list of Argentine ant immune genes, from which the 90 candidate genes were selected from
	Lhum_RefSeq.csv
		reference annotation of the Argentine ant genome
	nth_CS323.gtf
		GTF file output from 
	phenotypes_northland.csv
		information on samples	
	viljakainen_viruses_nthland.blastx.outfmt6_clean.csv
		BLASTx output against the Vijakainen et al. (2018) database
	virus_hits_nthland.csv
		list of hits extracted from the BLAST outputs
	viruses_nth_refseq.blastx.outfmt6_clean.csv
		BLASTx output against the viral nr database

02-scripts contains:

	Lester_et_al_2019.R
		Bash script used for RNA-seq reads processing
	Lester_et_al_2019.R
		R script including the analysis for both RNA-seq and RT-qPCR data, the two variables DataDir and FigDir have to be changed according to the user's file system

03-figures is an empty directory where figures generated by the R script are saved
